table/tables/mutation_checker.go

```diff
@@ -375,3 +380,58 @@ func getOrBuildColumnMaps(
 	}
 	return maps
 }
+
+// CheckIndexCount checks the number of row/index mutations before the txn commits,
+// to prevent some inconsistent transactions.
+//
+// Claim: For mutations in each table in the transaction, #(row insertions) != 0 => #(index insertions) % #(row insertions) = 0
+// Proof:
+// Without loss of generality, assume there is only 1 table involved in the transaction.
+// We examine the changes of numbers of these mutations: PI(put index), PR(put record), DI(delete index), DR(delete record),
+//  for each single-row operations. We argue that delta(PI) / delta(PR) = N always holds, where N is the number of indices.
+// Let h denote the handle, v denote the value of the row (or corresponding index mutations).
+// add record(h, v)
+//    h does not exist
+//        v no duplicates (index does not exist): PR + 1, PI + N
+//        v duplicated (0 puts, M dels): PR + 1, DI - M, PI + N
+// 			note: duplicated index mutations can only come from deletions on unique indices
+//    h is del
+//        v no duplicates: DR - 1, PR + 1, PI + N
+//        v duplicated (0 puts, M dels): DR - 1, PR + 1, DI - M, PI + N
+// update record (h, old, new),
+//    h is put (the intersection of old and new is M): DI + N - M
+//    h does not exist (the intersection of old and new is M): PR + 1, DI + N - M, PI + N
+// remove record(h)
+//    h is put: DR + 1, PR - 1, DI + N, PI - N
+//    h is del: unchanged
+//    h does not exist: DR + 1, DI + N
+func CheckIndexCount(txn kv.Transaction) error {
+	memBuffer := txn.GetMemBuffer()
+	if memBuffer == nil {
+		return nil
+	}
+
+	// count for each table
+	indexInsertionCount := make(map[int64]int)
+	rowInsertionCount := make(map[int64]int)
+	f := func(k kv.Key, v []byte) error {
+		if len(v) > 0 {
+			tableID := tablecodec.DecodeTableID(k)
+			if rowcodec.IsRowKey(k) {
+				rowInsertionCount[tableID] += 1
+			} else {
+				indexInsertionCount[tableID] += 1
+			}
+		}
+		return nil
+	}
+	if err := kv.WalkMemBuffer(memBuffer, f); err != nil {
```

Comment:
Does it process entries that's not modified but pessimistic-locked? Can it work with [this](https://github.com/pingcap/tidb/pull/25730) change?

Reply:
Yes it will process the locked but not changed keys, as each unique index key used to fetch rowkey will be turned into a `PUT` record, seems they are not compatibe, consider this:
```
create table t1(c1 int key, c2 int, c3 int, unique key uk(c2), key k1(c3));
insert into t1 values(1, 1, 1),(2, 2, 2),(3, 3, 3);
update t1 set c3 = c3 + 1 where c1 > 1;
here the to be committed keys are:
PUT uk 1->1 uk 2->2 uk 3->3
PUT sk  3_2 -> 2
PUT sk  4_3 -> 3
DEL sk   3_3 -> 3
DEL sk   2_2  -> 2
PUT rowkey  3-> 3, 3, 4
PUT rowkey  2-> 2, 2, 3
```

The number of `PUT` on row is 2 and the number of `PUT` on index is 5？

Reply:
@ekexium 
Seems @MyonKeminta has found a key point.

Reply:
Right.. it's a intentional special case. It seems to me we can't simply tell whether it comes from the optimization. A workaround might be adding a flag to indicate the special usage so they can be ignored in the check.

Reply:
The change is actually not expected and it's a temoprary solution to workaround the performance issues with many `LOCK` records in the `write cf`. Maybe we could put back this check after solving the `LOCK` record issue so that these tricky optimization or transformation could be removed.

Reply:
I'm thinking if there is a method to distinguish the usage and won't block removing it.

Reply:
If we have to add some new flags to membuffer to handle this, I would suspect whether it worths...